File: src/utils/demoData.ts
/**
 * Demo / Test Data for eStundnzettl
 * Generates realistic sample entries for 3 months with "Thomas Berger"
 */
import { WORK_CODE, WORK_MODELS } from "../hooks/constants";
import { toLocalDateString } from "../utils";
import { calculateEntryNetDuration } from "./timeCalculations";
import type { Entry, WorkCode } from '../types';

interface DemoUser {
  name: string;
  position: string;
  company: string;
  workDays: number[];
  workModelId: string;
  photo: string | null;
}

const DEMO_USER: DemoUser = {
  name: "Thomas Berger",
  position: "Monteur",
  company: "Berger Elektrotechnik GmbH",
  workDays: [...WORK_MODELS[0].days], // 38,5h Standard: Mo-Do 8,5h / Fr 4,5h
  workModelId: "38.5-classic",
  photo: null,
};

const DEMO_WORK_CODES: WorkCode[] = [
  { id: 1, label: "01 - Montage" },
  { id: 2, label: "02 - Büro" },
  { id: 3, label: "03 - Besprechung" },
  { id: 4, label: "04 - Material" },
  { id: 5, label: "05 - Wartung" },
  { id: 6, label: "06 - Reparatur" },
  { id: WORK_CODE.DRIVE, label: "19 - Fahrzeit" },
  { id: WORK_CODE.ARRIVAL, label: "19 - An/Abreise" },
  { id: WORK_CODE.OFFICE, label: "70 - Büro" },
];

// Generate a date string N days ago
const _daysAgo = (n: number): string => {
  const d = new Date();
  d.setDate(d.getDate() - n);
  return toLocalDateString(d);
};

// Generate work entry
const _workEntry = (date: string, start: string, end: string, pause: number, project: string, code: number): Entry => ({
  id: Date.now() + Math.random(),
  type: "work",
  date,
  start,
  end,
  pause,
  project,
  code,
  netDuration: calculateEntryNetDuration({
    entryType: "work",
    startTime: start,
    endTime: end,
    pauseDuration: pause,
    formDate: date,
    userData: null,
    code,
  }),
});

// Hilfsfunktion: Zufallsauswahl
const pick = <T>(arr: T[]): T => arr[Math.floor(Math.random() * arr.length)];

// Generate entries for the past 3 months
export const generateDemoEntries = (): Entry[] => {
  const entries: Entry[] = [];
  const today = new Date();
  let id = 1700000000000;

  const projects = [
    "BV Maier - Neubau",
    "BV Stadler - Sanierung",
    "BV Gemeinde Hof",
    "Lager / Werkstatt",
    "BV Kirchner - Umbau",
  ];
  const workCodes = [1, 3, 4, 5, 6];

  for (let dayOffset = 0; dayOffset < 90; dayOffset++) {
    const d = new Date(today);
    d.setDate(d.getDate() - dayOffset);
    // Nicht in der Zukunft
    if (d > today) continue;
    const dateStr = toLocalDateString(d);
    const dayOfWeek = d.getDay(); // 0=Sun, 6=Sat

    // Skip weekends
    if (dayOfWeek === 0 || dayOfWeek === 6) continue;

    // Random sick days (~2 per month)
    if (Math.random() < 0.07) {
      const target = dayOfWeek === 5 ? 270 : 510;
      entries.push({
        id: id++,
        type: "sick",
        date: dateStr,
        start: null,
        end: null,
        pause: 0,
        project: "Krank",
        code: null,
        netDuration: target,
      });
      continue;
    }

    // Random vacation days (~1 per month)
    if (Math.random() < 0.04) {
      const target = dayOfWeek === 5 ? 270 : 510;
      entries.push({
        id: id++,
        type: "vacation",
        date: dateStr,
        start: null,
        end: null,
        pause: 0,
        project: "Urlaub",
        code: null,
        netDuration: target,
      });
      continue;
    }

    // Freitag: kürzerer Tag (4,5h Soll)
    const isFriday = dayOfWeek === 5;
    const project = pick(projects);
    const code = pick(workCodes);

    let startH: number, startM: number, endH: number, endM: number, pause: number;

    if (isFriday) {
      startH = pick([6, 7, 7]);
      startM = pick([0, 0, 30]);
      endH = startH + pick([5, 5, 6]);
      endM = pick([0, 15, 30]);
      pause = pick([0, 15, 30]);
    } else {
      startH = pick([6, 6, 7, 7]);
      startM = pick([0, 0, 15, 30]);
      endH = startH + pick([9, 9, 10, 10]);
      endM = pick([0, 15, 30, 45]);
      pause = pick([30, 30, 30, 45]);
    }

    const start = `${String(startH).padStart(2, "0")}:${String(startM).padStart(2, "0")}`;
    const end = `${String(Math.min(endH, 20)).padStart(2, "0")}:${String(endM).padStart(2, "0")}`;

    entries.push({
      id: id++,
      type: "work",
      date: dateStr,
      start,
      end,
      pause,
      project,
      code,
      netDuration: calculateEntryNetDuration({
        entryType: "work",
        startTime: start,
        endTime: end,
        pauseDuration: pause,
        formDate: dateStr,
        userData: DEMO_USER,
        code,
      }),
    });

    // ~30% der Tage: zusätzlicher Fahrzeit-Eintrag (An/Abreise)
    if (Math.random() < 0.3) {
      const driveStart = `${String(startH - 1).padStart(2, "0")}:00`;
      const driveEnd = `${String(startH).padStart(2, "0")}:00`;
      entries.push({
        id: id++,
        type: "work",
        date: dateStr,
        start: driveStart,
        end: driveEnd,
        pause: 0,
        project,
        code: WORK_CODE.ARRIVAL,
        netDuration: calculateEntryNetDuration({
          entryType: "work",
          startTime: driveStart,
          endTime: driveEnd,
          pauseDuration: 0,
          formDate: dateStr,
          userData: DEMO_USER,
          code: WORK_CODE.ARRIVAL,
        }),
      });
    }
  }

  return entries;
};

export const DEMO_DATA = {
  user: DEMO_USER,
  workCodes: DEMO_WORK_CODES,
  generateEntries: generateDemoEntries,
};

File: native/core/src/main/kotlin/com/estundnzettl/core/calc/DemoData.kt
package com.estundnzettl.core.calc

import com.estundnzettl.core.model.Entry
import com.estundnzettl.core.model.EntryId
import com.estundnzettl.core.model.EntryType
import com.estundnzettl.core.model.UserData
import com.estundnzettl.core.model.WORK_MODELS
import com.estundnzettl.core.model.WorkCode
import java.time.DayOfWeek
import java.time.LocalDate
import kotlin.random.Random

/**
 * Demo-/Testdaten — Port von src/utils/demoData.ts.
 * Erzeugt realistische Beispieleinträge für 3 Monate mit "Thomas Berger".
 */

val DEMO_USER = UserData(
    name = "Thomas Berger",
    position = "Monteur",
    company = "Musterbetrieb",
    workDays = WORK_MODELS[0].days, // 38,5h Standard: Mo–Do 8,5h / Fr 4,5h
    workModelId = "38.5-classic",
    photo = null,
)

val DEMO_WORK_CODES = listOf(
    WorkCode(1, "01 - Arbeit"),
    WorkCode(2, "02 - Büro"),
    WorkCode(3, "03 - Besprechung"),
    WorkCode(4, "04 - Vorbereitung"),
    WorkCode(5, "05 - Wartung"),
    WorkCode(6, "06 - Sonstiges"),
    WorkCode(WorkCodes.DRIVE, "19 - Fahrzeit"),
    WorkCode(WorkCodes.ARRIVAL, "19 - An/Abreise"),
    WorkCode(WorkCodes.OFFICE, "70 - Büro"),
)

private fun <T> pick(random: Random, items: List<T>): T = items[random.nextInt(items.size)]

private fun hhmm(h: Int, m: Int): String =
    "${h.toString().padStart(2, '0')}:${m.toString().padStart(2, '0')}"

/** Einträge der letzten 90 Tage (Wochentage, gelegentlich Krank/Urlaub/Fahrt). */
fun generateDemoEntries(
    today: LocalDate = LocalDate.now(),
    random: Random = Random.Default,
): List<Entry> {
    val entries = ArrayList<Entry>()
    var id = 1_700_000_000_000L

    val projects = listOf("Projekt A", "Projekt B", "Projekt Gemeinde", "Büro / Werkstatt", "Projekt C")
    val workCodes = listOf(1, 3, 4, 5, 6)

    for (dayOffset in 0 until 90) {
        val d = today.minusDays(dayOffset.toLong())
        val dateStr = d.toString()
        val dayOfWeek = d.dayOfWeek

        // Wochenenden überspringen
        if (dayOfWeek == DayOfWeek.SATURDAY || dayOfWeek == DayOfWeek.SUNDAY) continue

        val isFriday = dayOfWeek == DayOfWeek.FRIDAY

        // ~2 Krankenstände pro Monat
        if (random.nextDouble() < 0.07) {
            val target = if (isFriday) 270 else 510
            entries.add(
                Entry(
                    id = EntryId.of(id++), type = EntryType.SICK, date = dateStr,
                    start = null, end = null, pause = 0,
                    project = "Krank", code = null, netDuration = target,
                ),
            )
            continue
        }

        // ~1 Urlaubstag pro Monat
        if (random.nextDouble() < 0.04) {
            val target = if (isFriday) 270 else 510
            entries.add(
                Entry(
                    id = EntryId.of(id++), type = EntryType.VACATION, date = dateStr,
                    start = null, end = null, pause = 0,
                    project = "Urlaub", code = null, netDuration = target,
                ),
            )
            continue
        }

        val project = pick(random, projects)
        val code = pick(random, workCodes)

        val startH: Int
        val startM: Int
        val endH: Int
        val endM: Int
        val pause: Int
        if (isFriday) {
            startH = pick(random, listOf(6, 7, 7))
            startM = pick(random, listOf(0, 0, 30))
            endH = startH + pick(random, listOf(5, 5, 6))
            endM = pick(random, listOf(0, 15, 30))
            pause = pick(random, listOf(0, 15, 30))
        } else {
            startH = pick(random, listOf(6, 6, 7, 7))
            startM = pick(random, listOf(0, 0, 15, 30))
            endH = startH + pick(random, listOf(9, 9, 10, 10))
            endM = pick(random, listOf(0, 15, 30, 45))
            pause = pick(random, listOf(30, 30, 30, 45))
        }

        val start = hhmm(startH, startM)
        val end = hhmm(minOf(endH, 20), endM)

        entries.add(
            Entry(
                id = EntryId.of(id++), type = EntryType.WORK, date = dateStr,
                start = start, end = end, pause = pause,
                project = project, code = code,
                netDuration = calculateEntryNetDuration(
                    entryType = "work", startTime = start, endTime = end,
                    pauseDuration = pause, formDate = dateStr,
                    userData = DEMO_USER, code = code,
                ),
            ),
        )

        // ~30% der Tage: zusätzlicher Fahrzeit-Eintrag (An/Abreise)
        if (random.nextDouble() < 0.3) {
            val driveStart = hhmm(startH - 1, 0)
            val driveEnd = hhmm(startH, 0)
            entries.add(
                Entry(
                    id = EntryId.of(id++), type = EntryType.WORK, date = dateStr,
                    start = driveStart, end = driveEnd, pause = 0,
                    project = project, code = WorkCodes.ARRIVAL,
                    netDuration = calculateEntryNetDuration(
                        entryType = "work", startTime = driveStart, endTime = driveEnd,
                        pauseDuration = 0, formDate = dateStr,
                        userData = DEMO_USER, code = WorkCodes.ARRIVAL,
                    ),
                ),
            )
        }
    }

    return entries
}
